### ISSUE TITLE
Logout Issue (Invalid CSRF Token)

### ISSUE THREAD

**Issue by adhik** [ch4mpy/spring-addons#208]
Hi, thanks for your great works here.

I've tried following [your tutorial here](https://www.baeldung.com/spring-cloud-gateway-bff-oauth2) and so far it worked, except the logout attempt. It gave me `Invalid CSRF token` response in Chrome.

After debugging, I've found the issue here:

![image](https://github.com/ch4mpy/spring-addons/assets/8499271/dedde5b4-0915-491c-86e7-b0fbb256a1d7)

It seems the the CSRF token is inside `"Cookie"` and can't be retrived by using

```
exchange.getRequest().getHeaders().get(csrfToken.getHeaderName())
```

maybe it would work by doing this?

```
exchange.getRequest().getHeaders().getOrEmpty("Cookie").get(csrfToken.getHeaderName())
```

Other workaround in the frontend (Angular) perhaps to put the `X-XSRF-TOKEN` from Cookie into the custom header:

```
logout() {
    lastValueFrom(
      this.http.post(`${environment.baseApiUrl}/logout`, null, {
        headers: {
          'X-POST-LOGOUT-SUCCESS-URI': environment.appUrl,
          'X-XSRF-TOKEN': this.cookie.get('XSRF-TOKEN'),
        },
        observe: 'response',
      })
    )
      .then(resp => {
        const logoutUri = resp.headers.get('Location');
        if (!!logoutUri) {
          window.location.href = logoutUri;
        }
      })
      .finally(() => {
        this.refresh();
      });
  }
```

but I don't think it is a good idea :smile: 

Thanks.

---

My environment:

- JDK 17
- Spring boot version: 3.2.4
- Spring-addons version: 7.6.12
- Spring Cloud Gateway version: 2023.0.1

**Comment by ch4mpy** [comment]
The Logout works in all 3 applications (with CSRF). 

Angular auto adds an `X-XSRF-TOKEN` header when a CSRF cookie is accessible to Javascript (http-only false).

Maybe did you miss the part whith `csrf: cookie-accessible-from-js` in spring-addons client properties for the BFF?

**Comment by adhik** [comment]
Hi, so it turned out Angular issue since I used absolute URL for ex:

```
this.http.post(`http://localhost:7080/bff/logout`, null, {
....
```

Angular didn't include the XSRF-TOKEN from Cookie because of [this code here](https://github.com/angular/angular/blob/f84d7dde872bbb44fa7a482565c0046a06d7f55c/packages/common/http/src/xsrf.ts#L96). 

So the solution is to use relative URL like the one in your example:

```
this.http.post(`/bff/logout`, null, {
```

Thanks :+1:

**Comment by ch4mpy** [comment]
Yep, that limitation with absolute URLs and CSRF is something to keep in mind when working with Angular :/
